File: 10/main.go
package main

import (
	"bufio"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	// Читаем количество тестов
	line, _ := reader.ReadString('\n')
	t, _ := strconv.Atoi(strings.TrimSpace(line))

	// Используем scanner для чтения чисел
	scanner := bufio.NewScanner(reader)
	scanner.Split(bufio.ScanWords)

	readInt := func() int {
		scanner.Scan()
		v, _ := strconv.Atoi(scanner.Text())
		return v
	}

	for test := 0; test < t; test++ {
		// Читаем n и m
		n := readInt()
		m := readInt()

		// Читаем массивы a, b, c
		a := make([]int, m)
		b := make([]int, m)
		c := make([]int, m)

		for i := 0; i < m; i++ {
			a[i] = readInt()
		}
		for i := 0; i < m; i++ {
			b[i] = readInt()
		}
		for i := 0; i < m; i++ {
			c[i] = readInt()
		}

		edges := make([]Edge, m)
		for i := 0; i < m; i++ {
			edges[i] = Edge{a: a[i] - 1, b: b[i] - 1, w: c[i]} // 0-indexed
		}

		result := solve(n, edges)

		// Выводим результат
		sb := strings.Builder{}
		for i, v := range result {
			if i > 0 {
				sb.WriteByte(' ')
			}
			sb.WriteString(strconv.Itoa(v))
		}
		sb.WriteByte('\n')
		writer.WriteString(sb.String())
	}
}

type Edge struct {
	a, b, w int
}

// solve находит минимальный вес w для каждого размера компоненты k
func solve(n int, edges []Edge) []int {
	// Сортируем рёбра по весу
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].w < edges[j].w
	})

	// Инициализируем DSU
	parent := make([]int, n)
	rank := make([]int, n)
	size := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
		size[i] = 1
	}

	// result[k] = минимальный вес для размера k+1
	result := make([]int, n)
	for i := 0; i < n; i++ {
		result[i] = -1
	}

	// k=1: любая вершина достижима сама из себя с w=0
	result[0] = 0

	// maxReached отслеживает максимальный размер, для которого уже найден ответ
	maxReached := 1

	// Обрабатываем рёбра в порядке возрастания веса
	for _, e := range edges {
		// Пропускаем петли (они не меняют связность)
		if e.a == e.b {
			continue
		}

		rootA := find(parent, e.a)
		rootB := find(parent, e.b)

		if rootA != rootB {
			// Объединяем компоненты
			newSize := size[rootA] + size[rootB]
			union(parent, rank, size, rootA, rootB)

			// Обновляем результат для всех новых размеров
			for k := maxReached + 1; k <= newSize; k++ {
				result[k-1] = e.w
			}
			if newSize > maxReached {
				maxReached = newSize
			}
		}

		// Если достигли максимального размера, можно остановиться
		if maxReached == n {
			break
		}
	}

	return result
}

func find(parent []int, x int) int {
	if parent[x] != x {
		parent[x] = find(parent, parent[x])
	}
	return parent[x]
}

func union(parent, rank, size []int, x, y int) {
	if rank[x] < rank[y] {
		parent[x] = y
		size[y] += size[x]
	} else if rank[x] > rank[y] {
		parent[y] = x
		size[x] += size[y]
	} else {
		parent[y] = x
		size[x] += size[y]
		rank[x]++
	}
}

File: 10/main.rs
use std::io::{self, BufRead, BufWriter, Write, Read};

struct Edge {
    a: usize,
    b: usize,
    w: i32,
}

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Читаем количество тестов
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let t: usize = line.trim().parse().unwrap();

    // Используем scanner для чтения чисел
    let mut tokens = Vec::new();
    let mut buffer = String::new();
    reader.read_to_string(&mut buffer).unwrap();
    
    for token in buffer.split_whitespace() {
        tokens.push(token.parse::<i32>().unwrap());
    }
    
    let mut token_idx = 0;
    let mut read_int = || {
        let val = tokens[token_idx];
        token_idx += 1;
        val
    };

    for _test in 0..t {
        // Читаем n и m
        let n = read_int() as usize;
        let m = read_int() as usize;

        // Читаем массивы a, b, c
        let mut a = vec![0; m];
        let mut b = vec![0; m];
        let mut c = vec![0; m];

        for i in 0..m {
            a[i] = read_int();
        }
        for i in 0..m {
            b[i] = read_int();
        }
        for i in 0..m {
            c[i] = read_int();
        }

        let mut edges = Vec::with_capacity(m);
        for i in 0..m {
            edges.push(Edge {
                a: (a[i] - 1) as usize,
                b: (b[i] - 1) as usize,
                w: c[i],
            });
        }

        let result = solve(n, &mut edges);

        // Выводим результат
        for (i, &v) in result.iter().enumerate() {
            if i > 0 {
                write!(writer, " ").unwrap();
            }
            write!(writer, "{}", v).unwrap();
        }
        writeln!(writer).unwrap();
    }
}

// solve находит минимальный вес w для каждого размера компоненты k
fn solve(n: usize, edges: &mut [Edge]) -> Vec<i32> {
    // Сортируем рёбра по весу
    edges.sort_by_key(|e| e.w);

    // Инициализируем DSU
    let mut parent: Vec<usize> = (0..n).collect();
    let mut rank = vec![0; n];
    let mut size = vec![1; n];

    // result[k] = минимальный вес для размера k+1
    let mut result = vec![-1; n];

    // k=1: любая вершина достижима сама из себя с w=0
    result[0] = 0;

    // maxReached отслеживает максимальный размер, для которого уже найден ответ
    let mut max_reached = 1;

    // Обрабатываем рёбра в порядке возрастания веса
    for e in edges.iter() {
        // Пропускаем петли (они не меняют связность)
        if e.a == e.b {
            continue;
        }

        let root_a = find(&mut parent, e.a);
        let root_b = find(&mut parent, e.b);

        if root_a != root_b {
            // Объединяем компоненты
            let new_size = size[root_a] + size[root_b];
            union(&mut parent, &mut rank, &mut size, root_a, root_b);

            // Обновляем результат для всех новых размеров
            for k in (max_reached + 1)..=new_size {
                result[k - 1] = e.w;
            }
            if new_size > max_reached {
                max_reached = new_size;
            }
        }

        // Если достигли максимального размера, можно остановиться
        if max_reached == n {
            break;
        }
    }

    result
}

fn find(parent: &mut [usize], x: usize) -> usize {
    if parent[x] != x {
        parent[x] = find(parent, parent[x]);
    }
    parent[x]
}

fn union(parent: &mut [usize], rank: &mut [usize], size: &mut [usize], x: usize, y: usize) {
    if rank[x] < rank[y] {
        parent[x] = y;
        size[y] += size[x];
    } else if rank[x] > rank[y] {
        parent[y] = x;
        size[x] += size[y];
    } else {
        parent[y] = x;
        size[x] += size[y];
        rank[x] += 1;
    }
}

